### DISCUSSION TITLE
Improving OCaml support

### DISCUSSION THREAD

**Discussion by QuentinRa** [giraud/reasonml-idea-plugin#349] [Feature request]
This is a follow up of #348 . I want to improve the User eXperience for OCaml users [more details later].

**Notes**

After reading the code, I learned that there are classes related to odoc documentation. Maybe this would be a feature in README.md?

**Comment by giraud** [comment]
documentation is available through classic Idea actions, like ctrl-q. no specific workflow involved, hence no specific entry in the readme, but it can be updated.

**Reply by QuentinRa** [reply]
My bad, I wasn't aware of this. I mean, if you didn't wrote this code, it would not be as it is now right? and I wasn't expecting it. I wasn't aware of the shortcut too 😔, thanks 👍.

Maybe it could be improved, so that if there is no local documentation, the documentation of the declaration is used instead? As we have in Java. Or is this not worth it?

![image](https://user-images.githubusercontent.com/54904135/133622216-73b1e757-60f0-49ee-b4ae-53b909bcde8a.png)

![image](https://user-images.githubusercontent.com/54904135/133623041-f4a6a911-0a12-4bec-ab6c-16dc6e0a2b56.png)

**Reply by giraud** [reply]
yes, please fill a bug

**Comment by QuentinRa** [comment]
To summarize, my main goal is to be able to create a project with the wizard, and run a .ml file without any prior configuration, on IntelliJ+Windows. As in PyCharm/PHPStorm, there isn't something to set the SDK, it would have to be set in the Settings. In both cases, it means that we can call "ocaml" and "ocamlc". Either OCaml is installed locally (option to download OCaml) or installed globally. This is the same as we have for Java.

![image](https://user-images.githubusercontent.com/54904135/133658594-63554e0d-17c7-43e4-90b7-6d00d6d1e0c9.png)

I will use "ocaml" and try to link it with REPL classes

![image](https://user-images.githubusercontent.com/54904135/133658706-9758ffaa-dff7-4467-934b-3a3f58c618aa.png)

And I will have to add another configuration that will run a .ml file.

From my current understanding, I will have a lot of blockers (as each one seems difficult)

* [ ] Download and install OCaml (I already have a portable SDK for 4.10.0)
* [ ] Add a new Compiler (ocamlc)
* [ ] Using ocaml with REPL classes
* [ ] Configuration for OCaml files

From what I understood, it would be better to test adding a new compiler first (to check that I can, etc.). Then I would move to other tasks.

**Reply by giraud** [reply]
IMO, the most difficult thing is to find how to not use SdkType anymore and still have all benefits, like automatic indexing and having the ability to add sources to the new java-less Sdk. Indexing is the most important thing in IntelliJ and getting it right is primordial. 

also I'd look into how to create custom extension points for the plugin. I don't thikn it's complex (some info here https://plugins.jetbrains.com/docs/intellij/plugin-extension-points.html) and it'll help to define a plugin based on this one. that would help to define custom/different plugins based on this one.

**Reply by QuentinRa** [reply]
I'm not sure if we need to get rid of SdkType (I don't know if this is really something only available only in IntelliJ), but I'm planning to put OCaml binaries (and ocaml/ocamlc) here

![image](https://user-images.githubusercontent.com/54904135/133798981-9a184611-d5ca-44dc-88cb-d880ac899b99.png)

![image](https://user-images.githubusercontent.com/54904135/133799765-0f59d208-e468-4ca8-9533-06b11db64e68.png)

Since "libraries" are also available in Java, maybe we got something that we could use here? (I don't know yet, I'm still exploring)

<hr>

Extension points seem interesting. I need to learn more about the structure of the code, so I don't know yet if I could use it, I will keep it in mind, thanks! This look quite easy, but I'm tackling something else for now.

**Reply by QuentinRa** [reply]
On Windows, I could share a zip with OCaml. A user could open it and run ocaml.exe, or ocamlc.exe and use them just fine. But, this is a least 300Mb, more if they add extensions (ocamlformat, ...), so distributing of zip for each version and the extensions would not be a good idea. But, opam is handling that just fine. If I managed to download a portable/light version of opam (~45Mb) and use it to install any version of OCaml/extensions, then this would help me proposing this menu

![image](https://user-images.githubusercontent.com/54904135/133803675-6d5f783d-6fdc-4a61-a53a-fcac36431188.png)

And I could add later a page in the settings to add/update extensions, as we have in Node/npm

![image](https://user-images.githubusercontent.com/54904135/133801410-6e167870-f1e8-469f-8a9d-aae108ec04a8.png)

Now, the real problem is using this zip (ocamlc, ocaml, ...) to replace SdkType. First, I will try to add the zip in external librairies, then I will try to see how this is working, and what can I do.

**Reply by QuentinRa** [reply]
I would also have to think about Linux and MacOS users. Fortunately, this can use the "add jdk" or check "selected JDK" as this is easier on Linux/MacOS to install OCaml.

**Comment by giraud** [comment]
though about a repl, and what is implemented is not good at all. I like how it's done in cursive (https://cursive-ide.com/userguide/repl.html#interaction-with-the-editor) where you can have a file opened and you can just send highlited code to the repl.

**Reply by QuentinRa** [reply]
Yeah! This is one of my goals, and that's why I want the "ocaml" command. I want to make something like the R plugin in fact.

**Comment by QuentinRa** [comment]
I'm following your advice and trying to use extensions. Hope this will work 🚀, as I'm not able to test on PyCharm/PHPStorm (simply changing IU to PY to use PyCharm is not working, I will test each feature on a new empty plugin)

Inside IntelliJ, this is like this (empty for now)

![image](https://user-images.githubusercontent.com/54904135/133858031-5b2de6fc-aa14-4ce2-8c0b-827f13a91599.png)

Another plugin inside **PyCharm**

![image](https://user-images.githubusercontent.com/54904135/133859013-f8648263-892e-41b3-91ee-91e107496679.png)

inside **PHPStorm**

![image](https://user-images.githubusercontent.com/54904135/133859083-9003a57c-7ea0-4056-bc60-868e77682aca.png)

The plugin

https://github.com/casteng/i-pascal

Other notes,

* https://github.com/Hannah-Sten/TeXiFy-IDEA/blob/4bb48d7c18ccf73f359ea039699fcbe5713bf06e/src/nl/hannahsten/texifyidea/settings/sdk/LatexProjectSdkSetupValidator.kt
* https://github.com/casteng/i-pascal/blob/2fc46aa6468972599628e46db3bfe4f0a1a28837/plugin/src/sdk/PascalProjectSdkValidator.java

I think I will explore what casteng did, It may greatly help this plugin.

**Comment by QuentinRa** [comment]
Currently, I managed to make the SDK-related stuff work. I can 

* Detect the installed SDK (I'm looking in C:/ocaml, C:/OCaml64, and .opam/home/xxx/.opam or .opam/usr/xxxx/), I will add more later (/usr/bin, etc. or make something using the command whereis seem better), and make something cleaner

![image](https://user-images.githubusercontent.com/54904135/133909167-e8dd961d-69c2-48c2-8d7e-22faf83f69b0.png)

* Select an SDK (or add one manually)

![image](https://user-images.githubusercontent.com/54904135/133909177-a235ece9-92fc-4ebc-a968-0e3bef2e2962.png)

![image](https://user-images.githubusercontent.com/54904135/133909181-c285db41-37ce-4c1b-8891-d30a7cd2f7ea.png)

* The SDK is automatically loaded in "External libraries"

![image](https://user-images.githubusercontent.com/54904135/133909185-18a86b31-f8ae-47df-b5f6-34293af05fcc.png)

![image](https://user-images.githubusercontent.com/54904135/133909187-e870b43f-92a9-4934-ac7a-97b937d260b1.png)

It took me a day 😭. I started with your project after removing almost everything (parsers, ...), and I added things little by little, as I'm just testing things.

I have some school work, so I don't think I will be back before next week, I will try to push a mock repository after cleaning the code a bit.

<hr>

Next?

* I will clean the code ("my playground") and put it on GitHub (will take time)
* manually download the sources if the user did not have them (easy => GitHub)
* hide the little button to switch the JDK if the user isn't inside a .ml/.mli
* test in PyCharm/PHPStorm (I will have to find a way, and this may take a lot of time 😱)
* deal with the create/import wizard
* test on Linux

**Comment by giraud** [comment]
Hi, 

I have worked on the sdk problem in #365 .

facet is gone, only settings are used to select the switch:
![sdk01](https://user-images.githubusercontent.com/239612/144897069-3017a149-8a13-4a9d-aa2f-0be1460af37b.png)

it then generates an external library:
![sdk02](https://user-images.githubusercontent.com/239612/144897102-e9a7144f-3fb2-4fec-95dd-c7d62b14e1e2.png)

and that library is automatically added to the module to get indexation:
![sdk03](https://user-images.githubusercontent.com/239612/144897169-5aca17d6-57b5-47aa-a574-684abbed91c8.png)

Modules are not visible in minor IDE but they are still present. 
It will be possible to use OCaml outside IntelliJ.

**Reply by QuentinRa** [reply]
Hi! That's pretty nice! I like the way you tackled that 👍 .
I have some time now, so I hope I will be able to work on it too!
